File: app/src/main/java/dev/ragnarok/fenrir/adapter/GiftAdapter.java
package dev.ragnarok.fenrir.adapter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import dev.ragnarok.fenrir.Constants;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.model.Gift;
import dev.ragnarok.fenrir.picasso.PicassoInstance;
import dev.ragnarok.fenrir.util.AppTextUtils;
import dev.ragnarok.fenrir.util.Utils;
import dev.ragnarok.fenrir.util.ViewUtils;

public class GiftAdapter extends RecyclerView.Adapter<GiftAdapter.Holder> {

    private final Context context;
    private List<Gift> data;
    private ClickListener clickListener;

    public GiftAdapter(List<Gift> data, Context context) {
        this.data = data;
        this.context = context;
    }

    @NonNull
    @Override
    public Holder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new Holder(LayoutInflater.from(context).inflate(R.layout.item_gift, parent, false));
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(@NonNull Holder holder, int position) {
        Gift gift = data.get(position);
        if (!Utils.isEmpty(gift.getThumb())) {
            ViewUtils.displayAvatar(holder.thumb, null, gift.getThumb(), Constants.PICASSO_TAG);
        } else {
            PicassoInstance.with().cancelRequest(holder.thumb);
        }
        if (!Utils.isEmpty(gift.getThumb())) {
            holder.message.setVisibility(View.VISIBLE);
            holder.message.setText(gift.getMessage());
        } else {
            holder.message.setVisibility(View.GONE);
        }
        if (gift.getDate() == 0)
            holder.time.setVisibility(View.GONE);
        else {
            holder.time.setVisibility(View.VISIBLE);
            holder.time.setText(AppTextUtils.getDateFromUnixTime(context, gift.getDate()));
        }
        holder.gift_container.setOnClickListener(v -> {
            if (clickListener != null) {
                clickListener.onOpenClick(holder.getBindingAdapterPosition(), gift);
            }
        });
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public void setData(List<Gift> data) {
        this.data = data;
        notifyDataSetChanged();
    }

    public void setClickListener(ClickListener clickListener) {
        this.clickListener = clickListener;
    }

    public interface ClickListener {
        void onOpenClick(int index, Gift gift);
    }

    public static class Holder extends RecyclerView.ViewHolder {

        final ImageView thumb;
        final TextView message;
        final TextView time;
        final View gift_container;

        public Holder(View itemView) {
            super(itemView);
            thumb = itemView.findViewById(R.id.item_thumb);
            message = itemView.findViewById(R.id.item_message);
            time = itemView.findViewById(R.id.item_time);
            gift_container = itemView.findViewById(R.id.gift_container);
        }
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/adapter/GiftAdapter.kt
package dev.ragnarok.fenrir.adapter

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import dev.ragnarok.fenrir.Constants
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.model.Gift
import dev.ragnarok.fenrir.nonNullNoEmpty
import dev.ragnarok.fenrir.picasso.PicassoInstance.Companion.with
import dev.ragnarok.fenrir.util.AppTextUtils
import dev.ragnarok.fenrir.util.ViewUtils.displayAvatar

class GiftAdapter(private var data: List<Gift>, private val context: Context) :
    RecyclerView.Adapter<GiftAdapter.Holder>() {
    private var clickListener: ClickListener? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder {
        return Holder(LayoutInflater.from(context).inflate(R.layout.item_gift, parent, false))
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: Holder, position: Int) {
        val gift = data[position]
        if (gift.thumb.nonNullNoEmpty()) {
            displayAvatar(holder.thumb, null, gift.thumb, Constants.PICASSO_TAG)
        } else {
            with().cancelRequest(holder.thumb)
        }
        if (gift.thumb.nonNullNoEmpty()) {
            holder.message.visibility = View.VISIBLE
            holder.message.text = gift.message
        } else {
            holder.message.visibility = View.GONE
        }
        if (gift.date == 0L) holder.time.visibility = View.GONE else {
            holder.time.visibility = View.VISIBLE
            holder.time.text = AppTextUtils.getDateFromUnixTime(context, gift.date)
        }
        holder.gift_container.setOnClickListener {
            clickListener?.onOpenClick(holder.bindingAdapterPosition, gift)
        }
    }

    override fun getItemCount(): Int {
        return data.size
    }

    fun setData(data: List<Gift>) {
        this.data = data
        notifyDataSetChanged()
    }

    fun setClickListener(clickListener: ClickListener?) {
        this.clickListener = clickListener
    }

    interface ClickListener {
        fun onOpenClick(index: Int, gift: Gift)
    }

    class Holder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val thumb: ImageView = itemView.findViewById(R.id.item_thumb)
        val message: TextView = itemView.findViewById(R.id.item_message)
        val time: TextView = itemView.findViewById(R.id.item_time)
        val gift_container: View = itemView.findViewById(R.id.gift_container)
    }
}